### ISSUE TITLE
Need have a way to enforce the existence of IOMMU protocol. (Bugzilla Bug 2071)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#9773]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 2071
Date: 2019-08-09T10:14:50+00:00
From: @jyao1
To:   unassigned &lt;<unassigned>&gt;
CC:   @corthon, nobody, @spbrogan, @lzeng14

Last updated: 2022-02-15T21:30:19+00:00

**Comment by tianocore-issues** [comment]
## Comment 8912
Date: 2019-08-09 10:14:50 +0000
From: @jyao1

- Industry Specification: ---
- Release Observed: EDK II Master
- Releases to Fix: EDK II Master
- Target OS: ---
- Bugzilla Assignee(s): unassigned &lt;<unassigned>&gt;

Since IOMMU is important, we need add a PCD (gEfiMdeModulePkgTokenSpaceGuid.PcdRequireIommu) and support code to enable a platform to require that the IOMMU be present before allowing PCI DMA operations.

In map/unmap/allocatebuffer/freebuffer, if PcdRequireIommu is true and IOMMU is not present, error should be returned.

**Comment by tianocore-issues** [comment]
## Comment 8922
Date: 2019-08-10 02:34:02 +0000
From: @jyao1

The reference can be found at https://github.com/microsoft/mu_basecore/tree/dev/201905/MdeModulePkg/Bus/Pci/PciHostBridgeDxe

**Comment by tianocore-issues** [comment]
## Comment 9097
Date: 2019-08-20 22:18:32 +0000
From: @lzeng14

Personally, it seems a pure platform policy, platform can do a check like below right before connecting root bridge in platform BDS?

    Status = gBS-&gt;LocateProtocol (
          &amp;gEdkiiIoMmuProtocolGuid,
          NULL,
          (VOID **) &amp;IoMmuProtocol
          );
    ASSERT_EFI_ERROR (Status);

**Comment by tianocore-issues** [comment]
## Comment 9212
Date: 2019-08-23 11:29:44 +0000
From: @lersek

In OvmfPkg, we had developed the following mechanism:

* rely on a particular platform module (typically the IOMMU DXE driver)
  to dynamically install one of two protocol GUIDs in the UEFI protocol
  database, namely:
  - gIoMmuAbsentProtocolGuid (&quot;IOMMU absent&quot;), or
  - gEdkiiIoMmuProtocolGuid (actual IOMMU protocol implementation)

* implement an empty library instance, to be used only via NULL class
  resolution, such that it have an empty CONSTRUCTOR function, and a
  DEPEX on (gEdkiiIoMmuProtocolGuid OR gIoMmuAbsentProtocolGuid)

* in the platform DSC file, add the IOMMU DXE driver to the platform

* in the platform DSC file, hook the library instance, via NULL class
  resolution, into *all* such edk2 DXE drivers that *optionally* consume
  gEdkiiIoMmuProtocolGuid -- that is, drivers that use the IOMMU if the
  IOMMU is available, and work without the IOMMU otherwise.

The end result is that the dispatching of the relevant universal edk2
DXE drivers is *delayed* until the platform IOMMU DXE driver makes a
decision. Once the decision is made (and either one of both protocol
GUIDs is installed), the DXE Core is free to dispatch the universal
drivers. And those drivers will consume, or not consume, the IOMMU in
accordance with the platform IOMMU DXE driver&apos;s decision.

In OVMF this was implemented for the AMD SEV (Secure Encrypted
Virtualization) IOMMU, in the following commits:

- c5d736eb6770 (&quot;OvmfPkg: Introduce IoMmuAbsent Protocol GUID&quot;,
  2017-07-10)
- 6c22e534beaa (&quot;OvmfPkg: Add PlatformHasIoMmuLib&quot;, 2017-07-10)
- f9d129e68a45 (&quot;OvmfPkg: Add IoMmuDxe driver&quot;, 2017-07-10)
- c6ab9aecb71b (&quot;OvmfPkg: update PciHostBridgeDxe to use
  PlatformHasIoMmuLib&quot;, 2017-07-10)

I think the above mechanism, already present in edk2, can be used for
the requirement reported in comment#0. Because, with none of the
relevant protocol GUIDs present in the protocol database, the PCI host
bridge / root bridge driver will never be dispatched. In turn, PciIo
protocol instances will not be produced by PciBusDxe, and then PCI
device drivers cannot bind PCI controllers. Hence they cannot set up DMA
either.

Basically, the idea is to move &quot;gIoMmuAbsentProtocolGuid&quot; and
&quot;PlatformHasIoMmuLib&quot; to MdeModulePkg, from OvmfPkg. Then other
platforms can use them too.

Anyway, my request is that, whatever IOMMU enforcement is implemented,
please keep the above construct functional. Thanks.

**Comment by tianocore-issues** [comment]
## Comment 9214
Date: 2019-08-23 12:18:26 +0000
From: @spbrogan

I like the different approaches.  Let me add a few comments to see if we can keep the discussion going.

@Laszlo - I like it and we use patterns like this when we are adding an optional feature and want to do it in such a way to not require common code changes but i don&apos;t think it is best for this use case.
  
1. This is a critical security feature and moving forward should default to on.  To me this means we want the simplistic path to have it on and require usage of IOMMU.  If a product wants to lower the security posture of their platform or must because of hardware design they have to make a configuration change to turn it off. 

2. It is too easy for a developer to miss adding the DEPEX to a module.  New code gets added or integrated and the library class is missed.  In the workflow and churn of commercial  PC products there is too much room for error and not a good &quot;audit&quot; story as it will work fine without the protocol.  
 
@Star - I like how simple your proposal is but again due to the fact that nearly all platforms customize their BDS this change will be hard to get and maintain consistency. 

Another solution would be to remove the logic for map/unmap/allocate/free from the PCI host bridge driver and always defer to a protocol.  The platform could then choose which protocol to publish based on the platform design. This would require a change to the OvmfPkg design.  


I am working under the assumption:
* We all want to start moving EDK2 to a more secure by default design.  
* We all agree that for most products built from Edk2 we should encourage IOMMU protections.  

If that isn&apos;t yet an agreed position then I would like to understand that first.

**Comment by tianocore-issues** [comment]
## Comment 9234
Date: 2019-08-25 23:42:03 +0000
From: @lzeng14

Seemingly, the flow in OvmfPkg is about enforcing DMA operation must be dispatched after IOMMU driver makes the decision (present or absent). The requirement of this BZ is about enforcing IOMMU present.

@Laszlo, since IoMmuDxe is a dxe driver, it will normally be dispatched before uefi device driver to do DMA operation, Is there some special case making the dependency mechanism must be used?

@Sean, using PCD also requires platforms&apos; customization (setting the PCD to TRUE/FALSE). A point is that could the default value be TRUE at current situation (need much platforms to be updated) ?

**Comment by tianocore-issues** [comment]
## Comment 9246
Date: 2019-08-26 10:46:29 +0000
From: @lersek

I agree that the present solution (protocol notify) in PciHostBridgeDxe ensures that, if the DXE phase dispatches an IOMMU driver, then PciHostBridgeDxe will take the IOMMU into use, before any UEFI driver can attempt DMA (in BDS). However, that looked like an implementation detail in PciHostBridgeDxe, and I didn&apos;t want to rely on it. Relying on a PI-based sequencing mechanism (depex) looks more robust.

I&apos;m OK if we introduce a FeaturePCD (default value TRUE), for requiring an IOMMU. That doesn&apos;t seem to break the current OvmfPkg usage pattern. It&apos;s just that, in the patch set, please include such OvmfPkg and ArmVirtPkg patches then that flip the PCD value to FALSE, in the corresponding DSC files. It&apos;s OK to change default behavior if platform compatibility is preserved in the same patch series, for in-tree code.

I do realize that this would be a larger hurdle for DSC files that live in the edk2-platforms repository. It&apos;s not impossible to solve, however.

- First, introduce the new Feature PCD to the DEC file, with default value FALSE. Also, implement the PCD-dependent logic in the PCI infrastructure. Wait until this patch is merged in edk2.

- Second, post a patch series for edk2-platforms that sets the PCD explicitly to FALSE in all DSC files in the tree. Wait until this is merged.

- Third, post a patch series for edk2 that (a) explicitly sets the PCD to FALSE in OvmfPkg and ArmVirtPkg DSC files, (b) flips the default value to TRUE in the DEC file.

**Comment by tianocore-issues** [comment]
## Comment 9247
Date: 2019-08-26 11:03:50 +0000
From: @lzeng14

Still think it will be better to keep the compatibility for open and internal platforms, and let platform (maybe platform BDS) to make the enforcement.
